File: tests/python3/src/T004_GmsecException.py
#!/usr/bin/env python3

import libgmsec_python3 as lp
from Test import Test


# Note: This module is named T004_GmsecException to ensure conformity with
#       unit tests from other language bindings.
#
#       The tests below verify GmsecError.
#


class Test_GmsecError(Test):

    def test(self):
        self.test_constructor_1()
        self.test_constructor_2()
        self.test_constructor_3()
        self.test_copy_constructor()
        # Note: Accessors are tested in each test above.


    def test_constructor_1(self):
        lp.log_info("Test constructor (3 args)")

        emsg1 = "Something bad happened"

        e1 = lp.GmsecError(1, 2, emsg1)

        self.check("Unexpected error classification", e1.get_error_class() == 1)
        self.check("Unexpected error code", e1.get_error_code() == 2)
        self.check("Unexpected custom code", e1.get_custom_code() == 0)
        self.check("Unexpected error message", e1.get_error_message() == emsg1)
        self.check("Unexpected exception message", str(e1) == "[1,2,0] : " + emsg1)


    def test_constructor_2(self):
        lp.log_info("Test constructor (4 args)")

        emsg1 = "Something bad happened"

        e1 = lp.GmsecError(1, 2, 3, emsg1)

        self.check("Unexpected error classification", e1.get_error_class() == 1)
        self.check("Unexpected error code", e1.get_error_code() == 2)
        self.check("Unexpected custom code", e1.get_custom_code() == 3)
        self.check("Unexpected error message", e1.get_error_message() == emsg1)
        self.check("Unexpected exception message", str(e1) == "[1,2,3] : " + emsg1)


    def test_constructor_3(self):
        lp.log_info("Test constructor (Status arg)")

        reason = "Some worthy status"
        status = lp.Status(1, 2, reason, 3)

        e1 = lp.GmsecError(status)

        self.check("Unexpected error classification", e1.get_error_class() == 1)
        self.check("Unexpected error code", e1.get_error_code() == 2)
        self.check("Unexpected custom code", e1.get_custom_code() == 3)
        self.check("Unexpected error message", e1.get_error_message() == reason)
        self.check("Unexpected exception message", str(e1) == "[1,2,3] : " + reason)


    def test_copy_constructor(self):
        lp.log_info("Test copy-constructor")

        emsg = "Some worthy exception"

        e1 = lp.GmsecError(1, 2, 3, emsg)
        e2 = lp.GmsecError(e1)

        self.check("Unexpected error classification", e1.get_error_class() == e2.get_error_class())
        self.check("Unexpected error code", e1.get_error_code() == e2.get_error_code())
        self.check("Unexpected custom code", e1.get_custom_code() == e2.get_custom_code())
        self.check("Unexpected error message", e1.get_error_message() == e2.get_error_message())
        self.check("Unexpected exception message", str(e1) == str(e2))


Test.run('Test GmsecError', Test_GmsecError())

File: tests/cpp/src/T004_GmsecException.cpp
#include "TestCase.h"

using namespace gmsec::api5;


void test_constructor_1(Test& test)
{
	GMSEC_INFO << "Testing GmsecException constructor (3 args)...";

	std::string emsg1 = "Something bad happened";
	std::string emsg2;

	GmsecException e1(1, 2, emsg1.c_str());
	GmsecException e2(3, 4, NULL);

	test.check("Unexpected error class", 1 == e1.getErrorClass());
	test.check("Unexpected error code", 2 == e1.getErrorCode());
	test.check("Unexpected error custom code", 0 == e1.getCustomCode());
	test.check("Unexpected error message", emsg1 == e1.getErrorMessage());

	test.check("Unexpected error class", 3 == e2.getErrorClass());
	test.check("Unexpected error code", 4 == e2.getErrorCode());
	test.check("Unexpected error custom code", 0 == e2.getCustomCode());
	test.check("Unexpected error message", emsg2 == e2.getErrorMessage());
}


void test_constructor_2(Test& test)
{
	GMSEC_INFO << "Testing GmsecException constructor (4 args)...";

	std::string emsg1 = "Something bad happened";
	std::string emsg2;

	GmsecException e1(1, 2, 3, emsg1.c_str());
	GmsecException e2(4, 5, 6, NULL);

	test.check("Unexpected error class", 1 == e1.getErrorClass());
	test.check("Unexpected error code", 2 == e1.getErrorCode());
	test.check("Unexpected error custom code", 3 == e1.getCustomCode());
	test.check("Unexpected error message", emsg1 == e1.getErrorMessage());

	test.check("Unexpected error class", 4 == e2.getErrorClass());
	test.check("Unexpected error code", 5 == e2.getErrorCode());
	test.check("Unexpected error custom code", 6 == e2.getCustomCode());
	test.check("Unexpected error message", emsg2 == e2.getErrorMessage());
}


void test_constructor_3(Test& test)
{
	GMSEC_INFO << "Testing GmsecException constructor (w/ Status)...";

	std::string reason = "My status reason";

	Status status(1, 2, reason.c_str(), 3);

	GmsecException e1(status);

	test.check("Unexpected error class", 1 == e1.getErrorClass());
	test.check("Unexpected error code", 2 == e1.getErrorCode());
	test.check("Unexpected error custom code", 3 == e1.getCustomCode());
	test.check("Unexpected error message", reason == e1.getErrorMessage());
}


void test_copy_constructor(Test& test)
{
	GMSEC_INFO << "Testing GmsecException copy-constructor...";

	std::string emsg1 = "Something bad happened";

	GmsecException e1(1, 2, 3, emsg1.c_str());
	GmsecException e2(e1);

	test.check("Unexpected error class", e1.getErrorClass() == e2.getErrorClass());
	test.check("Unexpected error code", e1.getErrorCode() == e2.getErrorCode());
	test.check("Unexpected error custom code", e1.getCustomCode() == e2.getCustomCode());
	test.check("Unexpected error message", emsg1 == e2.getErrorMessage());
}


void test_assignment_operator(Test& test)
{
	GMSEC_INFO << "Testing GmsecException assignment-operator...";

	std::string emsg1 = "Something bad happened";

	GmsecException e1(1, 2, 3, emsg1.c_str());
	GmsecException e2(4, 5, 6, NULL);

	e2 = e1;

	test.check("Unexpected error class", e1.getErrorClass() == e2.getErrorClass());
	test.check("Unexpected error code", e1.getErrorCode() == e2.getErrorCode());
	test.check("Unexpected error custom code", e1.getCustomCode() == e2.getCustomCode());
	test.check("Unexpected error message", emsg1 == e2.getErrorMessage());
}


void test_what(Test& test)
{
	GMSEC_INFO << "Testing GmsecException what()...";

	std::string emsg1    = "Something bad happened";
	std::string expected = "[1,2,3] : Something bad happened";

	GmsecException e1(1, 2, 3, emsg1.c_str());

	test.check("Unexpected summarized error message", expected == e1.what());
}


int test_GmsecException(Test& test)
{
	test.setDescription("Test GmsecException");

	test_constructor_1(test);
	test_constructor_2(test);
	test_constructor_3(test);
	test_copy_constructor(test);
	test_assignment_operator(test);
	test_what(test);

	return 0;
}

TEST_DRIVER(test_GmsecException)
